HPCC-14111 To dfuplus, add capability to export and restore super files.

AttilaVamos opened the pull request:
Improve savexml action to generate proper logical and super file xml blob.

Improve add action to load simple logical and superfile metainformation
from an XML file. If the restored metainformation belong to a superfile
then update all sub-files SuperOwner tag.

Fix missing '\n' at the end of some DFUPlus return messages.

Signed-off-by: Attila Vamos attila.vamos@gmail.com

AttilaVamos commented:
@jakesmith please review.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

JamesDeFabia commented on dali/dfu/dfuutil.cpp:
```diff
@@ -868,10 +897,169 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            else
+            {
+                StringBuffer s;
+                throw MakeStringException(-1,"There is not Superfile name in XML");
```
Should this be 'There is no Superfile name in XML'

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -858,9 +858,38 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            CDfsLogicalFileName tmplfn;
+            StringBuffer lfnpath;
+            bool safe = false;
+            if ((strstr(lfn, "::") != NULL) && !strchr(lfn, '/'))
```
I am not sure why you are checking this, it looks like you are seeing whether this looks like a logical file or an xpath?
..it looks like you use it as an xpath if it doesn't look like a logical file below.

* If lfn was e.g. "myfile" - it would not have "::" in it, so it would try to connect to xpath "myfile".
Which is incorrect, previously it would lookup file "myfile" and succeed and use it to export.
* CDFUHelper::getFileXML() expected a logical file name or a group name before [the group bit is a bit bizarre, I don't think it's used anywhere].

It was using DFS interface to get XML, should continue to do, instead of going direct to Dali.
IDistributedFileDirectory::getFileTree() , can export the file xml, whether a plain file or super.

AttilaVamos replied:
Yes, the IDistributedFileDirectory::getFileTree() can export the file xml. The result is fine for plain file, but a bit  strange with superfile:
```xml
 <File directory="/var/lib/HPCCSystems/hpcc-data/thor" numclusters="1" numparts="6">
  <Attr fileCrc="-570320118"
        footerLength="0"
        headerLength="0"
        owner=""/>
  <Cluster defaultBaseDir="/var/lib/HPCCSystems/hpcc-data/thor"
           defaultReplicateDir="/var/lib/HPCCSystems/hpcc-mirror/thor"
           interleave="2"
           mapFlags="48">
   <Group>192.168.1.56</Group>
  </Cluster>
  <Part fileCrc="-1256801527"
        modified="2015-09-11T15:58:23"
        name="book2.csv._1_of_3"
        num="1"
        size="34"/>
  <Part fileCrc="0"
        modified="2015-09-09T14:36:24"
        name="book1.csv._1_of_3"
        num="2"
        size="0"/>
  <Part fileCrc="80761555"
        modified="2015-09-11T15:58:23"
        name="book2.csv._2_of_3"
        num="3"
        size="34"/>
  <Part fileCrc="0"
        modified="2015-09-09T14:36:24"
        name="book1.csv._2_of_3"
        num="4"
        size="0"/>
  <Part fileCrc="1051947210"
        modified="2015-09-11T15:58:23"
        name="book2.csv._3_of_3"
        num="5"
        size="34"/>
  <Part fileCrc="-570320118"
        modified="2015-09-09T14:36:24"
        name="book1.csv._3_of_3"
        num="6"
        size="102"/>
 </File>
 <SuperFile interleaved="2"
            modified="2015-09-14T09:08:05"
            name="superfile"
            numsubfiles="2">
  <SubFile name=".::book2.csv" num="1"/>
  <SubFile name=".::book1.csv" num="2"/>
  <Attr checkSum="2928410085"
        csvQuote="&quot;"
        csvSeparate="\,"
        csvTerminate="\n,\r\n"
        fileCrc="-570320118"
        footerLength="0"
        format="csv"
        headerLength="0"
        job="book2.csv"
        maxRecordSize="8192"
        owner=""
        recordSize="102"
        size="204"
        workunit="D20150911-165823">
   <ECL>RECORD
    STRING field1;
    STRING field2;
    STRING field3;
END;</ECL>
  </Attr>
  <OrigName>.::superfile</OrigName>
 </SuperFile>
```

jakesmith replied:
I may have been mistaken, but looking again at the server-side implementation of getFileTree() - it still looks like it will serialize the SuperFile meta only.

Can you check and debug what it's actually doing for a super file request?

If there is no way to do it, then a way via IDistributedFileDirectory should be added.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -868,10 +897,169 @@ class CDFUhelper: public CInterface, implements IDFUhelper
-        if (file)
-            file->attach(lfn,user);
+
+        // Check if this XML is a superfiel map
```
typo: superfile

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -868,10 +897,169 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+        else
+        {
+            // Logical file map
+            Owned<IRemoteConnection> conn = querySDS().connect(lfn, myProcessSession(), 0, FOREIGN_DALI_TIMEOUT);
```
Should not go direct to Dali, should use DFS methods, in this case IDistributedFileDirectory::lookup()

AttilaVamos replied:
Only a theoretical question: Why do not go direct to Dali?

richardkchapman replied:
Because we want to move this information to Cassandra sometime...

jakesmith replied:
various reasons, access to Dali (or e.g. Cassandra) needs to be tightly encapsulated, all access to it should go via well defined interfaces.
dfuplus is a tool that talks via those interfaces

AttilaVamos replied:
Thanks, it is clear now.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -868,10 +897,169 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                {
+                    queryDistributedFileDirectory().removeEntry(lfn,user);
+                    StringBuffer s;
+                    throw MakeStringException(-1,"Physical Part doesn't exists for logical file %s!", lfn);
```
>Physical Part

Physical parts ...

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -868,10 +897,169 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                    }
+                }
+            }
+            PROGLOG("%d of %d sub file(s) reference to superfile %s are fixed.",subnum, numSubfiles, superName.str());
```
There needs to be a similar approach to the handling of a standard file for super files.
IDFUhelper should not interact with Dali directly or deal with the inner workings of the file representation.
It use DFS methods to create objects (that validate if necessary).

There isn't a current [DFS] way to reattach a exported super file representation - but for this feature there should be.
The way I would do it, is to add a IDistributedFileDirectory::createNewSuper, which takes the IPropertyTree.
It can do some basic validation, then create a [unattached] CDistributedSuperFile / IDistributedSuperFile.
You can then use the existing methods to deal with SuperOwner etc., without going directly to the raw XML via IPT.

When validated, you can attach() it.

jakesmith commented:
@AttilaVamos - back to you

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith updated. Please re-review.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+    Owned<IDistributedSuperFile> superfile;
+
+    if (!tree->hasProp("@name"))
+        throw MakeStringException(-1, "createSuperFile: There is no SuperFile name in XML");
```
trivial: this test is unnecessary really, the test below would catch it and is sufficient.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+    // Get the correct branch name
+    StringBuffer branchName;
+    branchName.set(queryDfsXmlBranchName(DXB_SuperFile)).trim();
+    const char *branch = branchName.str();
```
trivial: may as well just branchName.str() below directly rather than declare *branch too.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+    sdlname.set(pSuperName);
+    StringBuffer head;
+    sdlname.makeScopeQuery(head.clear(), true);
+    Owned<IRemoteConnection> conn = querySDS().connect(head.str(), myProcessSession(), 0, FOREIGN_DALI_TIMEOUT);
```
This is going to fail if intermediate paths don't exist.
e.g. /Files/Scope[@name="scope1"]/Scope[@name="scope2"]/Scope[@name="scope3"]/SuperFile[@name="asf"]

Will not connect to unless /Files/Scope[@name="scope1"]/Scope[@name="scope2"]/Scope[@name="scope3"]/
is already present.

jakesmith replied:
More significantly, this isn't the approach we discussed.

Creating the IDistributedSuperFile should be akin to creating a IDistributeFile (see IDistributedFileDirectory::createNew). 
It should create and a IDistributedSuperFile, without committing changes to Dali.
Then you can do a super->attach() to attach it to Dali.
Verification of subfiles/consistentcy etc. should as much as possible use existing CDistributedSuperFile methods.

jakesmith replied:
re. intermediate scopes, see ensureFileScope()

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+            {
+                StringBuffer subquery;
+                sublname.makeFullnameQuery(subquery, DXB_File, true);
+                Owned<IRemoteConnection> subconn = querySDS().connect(subquery.str(), myProcessSession(), RTM_LOCK_WRITE, FOREIGN_DALI_TIMEOUT);
```
should use existing DFS methods as much as possible.
So you could 'lookup' here, rather than connect directly (which would bypass locking etc. etc.)

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+                subnum++;
+
+                Owned<IPropertyTree> subroot = subconn->getRoot();
+                Owned<IPropertyTreeIterator> iter = subroot->getElements("SuperOwner");
```
use existing methods.. in this case getSuperOwners()

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7930,130 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+    }
+    PROGLOG("%d of %d SubFile(s) reference to SuperFile '%s' are fixed.", subnum, numSubfiles, pSuperName);
+
+    return superfile.getLink();
```
I can't see where 'superfile' is set anywhere.. ?
See previous comment, should return a valid but unattached IDistributeSuperFile, which can then be attach()'d

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -579,6 +579,7 @@ interface IDistributedFileDirectory: extends IInterface
     virtual IFileDescriptor *getFileDescriptor(const char *lname,IUserDescriptor *user,const INode *foreigndali=NULL, unsigned foreigndalitimeout=FOREIGN_DALI_TIMEOUT) =0;
 
     virtual IDistributedSuperFile *createSuperFile(const char *logicalname,IUserDescriptor *user,bool interleaved,bool ifdoesnotexist=false,IDistributedFileTransaction *transaction=NULL) = 0;
+    virtual IDistributedSuperFile *createSuperFile(IPropertyTree *tree,IUserDescriptor *user,bool interleaved,bool ifdoesnotexist=false,IDistributedFileTransaction *transaction=NULL) = 0;
```
createNewSuper might be better name, akin to createNew and not to be confused with the existing 'createSuperFile' which creates an empty superfile that can be transactional.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,45 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+        if (t->hasProp("@numsubfiles"))
+        {
+            // It seems XML is fine to create  superfile
+            Owned<IDistributedSuperFile> superfile = queryDistributedFileDirectory().createSuperFile(t, user, false);
```
See previous comments, should create, then attach.
You should do the verification (e.g. subs exist) at attach time, since the IDistributeSuperFile object can exist/be passed around (unattached) and is valid until associated (attached) to Dali.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,45 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                if (file->existsPhysicalPartFiles(0))
+                {
+                    PROGLOG("Add logical file %s.", lfn);
+                    file->attach(lfn,user);
```
This call can become common, i.e. need to do same whether regular file or super.

jakesmith replied:
Except need (as with your current commit) to pass in a flag to indicate that it should load and link subfiles.
It should not load the subfiles at all at createNewSuperFile stage where the super is abstract and not attached.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,45 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc, true);
+            if (file)
+            {
+                if (file->existsPhysicalPartFiles(0))
```
Arguably this should be optional
There's no guarantee the physicals don't disappear after this command succeeds..
And there might be valid reasons for wanting to restore meta and physicals separately or not necessarily have to restore physicals ahead of meta.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,45 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                }
+                else
+                {
+                    queryDistributedFileDirectory().removeEntry(lfn, user);
```
Shouldn't be necessary .. file was never attach()'d

jakesmith commented:
@AttilaVamos - back to you.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@jakesmith Is this ready for you to re-review?

AttilaVamos commented:
@richardkchapman @jakesmith I don't finish the update yet. (I pushed it to sync between my desktop and laptop). It will be ready to new review by tomorrow.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith please review this attempt.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -999,6 +999,7 @@ protected: friend class CDistributedFile;
      */
     IDistributedFile *createNew(IFileDescriptor * fdesc, bool includeports=false);
     IDistributedSuperFile *createSuperFile(const char *logicalname,IUserDescriptor *user,bool interleaved,bool ifdoesnotexist,IDistributedFileTransaction *transaction=NULL);
+    IDistributedSuperFile *createNewSuperFile(ISuperFileDescriptor *desc, IUserDescriptor *user, bool includeports=false);
```
includeports not used, should be remove.

I noticed it's not used on createNew() either, so could be removed there too.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5225,12 +5226,38 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+    void reloadSubFiles()
+    {
+        loadSubFiles(NULL,INFINITE);
+    }
```
Shouldn't need these, work should be done in a CDistributedSuperFile constructor.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7956,14 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
 }
 
 // MORE: This should be implemented in DFSAccess later on
+IDistributedSuperFile *CDistributedFileDirectory::createNewSuperFile(ISuperFileDescriptor *desc, IUserDescriptor *user, bool includeports)
```
As when creating a IDistributeFile/createNew, shouldn't supply a IUserDescriptor this stage, as it is unattached.
A IUserDescriptor should only be needed when attach()'ing to DFS, at which point permissions are checked.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -7929,6 +7956,14 @@ IDistributedSuperFile *CDistributedFileDirectory::createSuperFile(const char *_l
+{
+    Owned<IDistributedSuperFile> superfile;
+    superfile.setown( new CDistributedSuperFile(this, desc, user));
+    return superfile.getLink();
```
no other paths and not doing anything else with object 'superfile', so simpler as:
```
return new CDistributedSuperFile(this, desc, user);
```

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -432,11 +432,14 @@ interface IDistributedSuperFile: extends IDistributedFile
                                 // returns file for part (not linked) NULL if not found
+
+    virtual void reloadSubFiles()=0;
+    virtual void relinkSubFile(unsigned pos)=0;
```
Not sure why you've extended the IDistributedSuperFile public interface..
Shouldn't need to, to construct a new CDistributedSuperFile object.

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -579,6 +582,7 @@ interface IDistributedFileDirectory: extends IInterface
     virtual IFileDescriptor *getFileDescriptor(const char *lname,IUserDescriptor *user,const INode *foreigndali=NULL, unsigned foreigndalitimeout=FOREIGN_DALI_TIMEOUT) =0;
 
     virtual IDistributedSuperFile *createSuperFile(const char *logicalname,IUserDescriptor *user,bool interleaved,bool ifdoesnotexist=false,IDistributedFileTransaction *transaction=NULL) = 0;
+    virtual IDistributedSuperFile *createNewSuperFile(ISuperFileDescriptor *desc, IUserDescriptor *user, bool includeports=false) = 0;
```
includeports unused.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user, NULL, FOREIGN_DALI_TIMEOUT, true, true);
```
>FOREIGN_DALI_TIMEOUT

clearer to go with the defaults here, as no foreign dali involved.
so clearer if change to:
```
Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user);
```

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+        if (t->hasProp("@numsubfiles"))
+        {
+            Owned<ISuperFileDescriptor> fdesc = deserializeSuperFileDescriptorTree(t);
+            // It seems XML is fine to create  superfile
```
The existing CSuperFileDescriptor(IPropertyTree *attr) constructor is not expecting the xml info for a whole superfile.
It expects the attributes to be given ("Attr") only, 

So this doesn't look right, though you may get away without error.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            // It seems XML is fine to create  superfile
+            Owned<IDistributedSuperFile> superfile = dfd.createNewSuperFile(fdesc, user, false);
+
+            if (superfile)
```
can never be NULL, so if (superfile) not needed.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                        if (!dfd.isSuperFile(subfilename.str(), user))
+                            if (!dfd.existsPhysical(subfilename.str(), user))
+                                throw MakeStringException(-1, "Physical Parts doesn't exists for logical file %s!", subfilename.str());
+                    }
```
the above validation code (which I think should probably be optional) would be better contained in a 'validate()' method, that may be useful for existing files/supers too.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+        else
+        {
+            // Logical file map
+            if (dfd.exists(lfn, user, true))
```
This should be done for both regular and super files

jakesmith replied:
This could move to before doing the Super or File specific code (if there is any) and pass in 'kind != DXB_SuperFile' to exists

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            Owned<IDistributedFile> file = dfd.createNew(fdesc, true);
+            if (file)
+            {
+                if (file->existsPhysicalPartFiles(0))
```
See comment re. IDistributedFile::validate() on super, std. file can perform this work there.

jakesmith commented on system/mp/mpbase.cpp:
```diff
@@ -556,7 +556,8 @@ protected: friend class CNodeIterator;
     {
         SocketEndpointArray epa;
-        epa.fromText(s,defport);
+        if (s)
```
Not really necessary now that your other change means SocketEndpointArray::fromText handles NULL.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5225,12 +5226,38 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+        parent = _parent;
+        setUserDescriptor(udesc, user);
+        interleaved=0;
+        root.set(_desc->getProperties());
```
getProperties() *should* always be returning "Attr" tree...
See other comment in dafdesc.cpp near call to deserializeSuperFileDescriptorTree()

So it's working by accident here at moment, you can't use getProperties() to get XML you need for this constructor..
And you can't use an existing ISuperFileDesciptor as a container for the raw XML, it wasn't intended for that.

From our IM conversation a while back:
```
Attila: How can I create a SuperFile from a SuperFileDescriptor? I  Can't find any constructor.

Jake: there isn't one.
existing createNew() is based on a IFileDescriptor, you could query the IFileDescriptor to see if it's a super IFileDescriptor::querySuperFileDescriptor()
but there is more work to do because ISuperFileDescriptor doesn't supprted all the meta data being imported.
i.e. you would 1st need to create a ISuperFileDescriptor from the import XML that fully represents it, then do createNew()
That would be most 'complete' way of doing it, but means extends (and potentially introducing problems) into the the implementation of ISuperFileDescriptor/CSuperFileDescriptor
You could look at it though and see how much work it would be / how disruptive it would be.
But what I suggest is that you skip creation of a 'full' ISuperFileDescriptor
and introduce a createNewSuper(IPropertyTree *super)
that creates the IDistributeSuperFile you really want directly.

Attila : with CDistributedSuperFile(CDistributedFileDirectory *_parent, IPropertyTree _root,const CDfsLogicalFileName &_name,IUserDescriptor user) constructor?

Jake: yes I think so, but don't think it needs name or user..
can't remember, might be supplied at attach time, let me look Sep 25, 2015 12:42
as until attached, it's  standalone, needs no authority(user)
yes attach takes user name, which makes sense
ah you were talking about an existing constructor, not a new one...
let me look at it, see what it does
ok, I see (to some extent) it needs user, because it looks up subfiles...
this is a bit different, it is creating a CDistributedSuperFile based on an existing published file.
may be very similar though
the only difference may be, that the name shouldn't be supplied.
does that make sense?
It should probably not load the subfiles upfront either.. loading means locking..
which would mean an unattached super, locks published subfiles, which would break the detach()'s semantics of a file.
So a new constructor:
CDistributedSuperFile(CDistributedFileDirectory *_parent, IPropertyTree _root);
which does similar init, except delays loadSubFiles to point where it's attach()'d
I'm sure I've glossed over a few details, but I think that's broadly correct
```


So as we discussed before, you should pass the IPropertyTree in directly to your new CDistributedSuperFile constructor.

Also mentioned above, the new constructor should be be very similar to the existing constructor
```
    CDistributedSuperFile(CDistributedFileDirectory *_parent, IPropertyTree *_root,const CDfsLogicalFileName &_name,IUserDescriptor* user) 
```
except:
1) You are creating an unattached file (not an object based on a published file), so no need for 'name' or 'user', that comes during attach()
2) You do not want the init to loadSubFiles, because that will lookup and lock them and you are creating an unattached file (no Dali connection at this point). It should delay that action until attach().

Your new constructor code should look something like:
```
    void commonInit(IPropertyTree *_root)
    {   
        root.set(_root);
        const char *val = root->queryProp("@interleaved");
        if (val&&isdigit(*val))
            interleaved = atoi(val);
        else
            interleaved = strToBool(val)?1:0;
    }

    CDistributedSuperFile(CDistributedFileDirectory *_parent, IPropertyTree *_root) // constructor for unattached superfile
    {
      commonInit(root);
    }
```

the existing init() can probably be changed to call commonInit() as doing common work.

The rest of the initialization should happen during attach() , i.e. the loadSubFiles and the linking.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                    superfile->reloadSubFiles();
+                    unsigned numSubFiles = superfile->numSubFiles();
+                    for (unsigned idx=0; idx < numSubFiles; idx++)
+                        superfile->relinkSubFile(idx);
```
See previous comments re. new constructor, it should not loadSubFiles(), but attach() should. 
loadSubFiles() _could_ (but doesn't) validate that the sub files belong to the super, so it would be better if you passed in an option to link them (to call linkSubFile() for the subfiles) as they load to be clear. (rather than do here)

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                {
+                    EXCLOG(e,"addFileXML(Super File");
+                    e->Release();
+                    dfd.removeSuperFile(lfn, true, user, 0);
```
With the above changes, either superfile->attach() succeeds or it will fire an exception, either way you will not need to remove here.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,83 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                    Owned<IPropertyTreeIterator> treeIter = t->getElements("SubFile");
+                    ForEach(*treeIter)
+                    {
+                        IPropertyTree & st = treeIter->get();
```
this will leak.

jakesmith commented:
@AttilaVamos  - back to you.

AttilaVamos commented:
@jakesmith please review this new attempt.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@AttilaVamos - am going to review changes now, however it's quite difficult to see what changes since the last review.
It would make it easier to review the new changes, if they were left as separate commits and only squashed when reviewing process is complete.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
-            file->attach(lfn,user);
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
+        // Check if this XML is a superfile map
+        if (t->hasProp("@numsubfiles"))
```
Not new to this round of changes I think, but a more obvious if was testing whether root node was 'SuperFile' instead here.

jakesmith replied:
It would be cleaner if did:
```
        DfsXmlBranchKind kind;
        if (0 == strcmp(t->queryName(), queryDfsXmlBranchName(DXB_SuperFile))
            kind = DXB_SuperFile;
        else if (0 == strcmp(t->queryName(), queryDfsXmlBranchName(DXB_File))
            kind = DXB_File;
        else
            throw MakeStringException(-1, "Unrecognised file XML root tag detected: '%s'", t->queryName());
```

Then use kind conditionally above where needed.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5882,6 +5900,28 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
 
+    void validate()
+    {
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
```
'parent' is guaranteed to be assigned by this stage. So should that.

jakesmith replied:
This comment hasn't been answered / address afaics.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5882,6 +5900,28 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+                if (!dfd.isSuperFile(subfilename.str(), NULL))
+                    if (!dfd.existsPhysical(subfilename.str(), NULL))
+                        throw MakeStringException(-1, "Physical Parts doesn't exists for logical file %s!", subfilename.str());
+            }
```
May be worth also validating that there number of "SubFile"'s equals numSubFiles here.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                }
+                else
+                {
+                    dfd.removeEntry(lfn, user);
```
It's not attached yet, so this is not required.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+        if (t->hasProp("@numsubfiles"))
+        {
+            if (dfd.exists(lfn, user))
+                throw MakeStringException(-1, "Super file '%s' already exists!", lfn);
```
This could move to before doing the Super or File specific code (if there is any) and pass in 'kind != DXB_SuperFile' to exists

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            // It seems XML is fine to create  super file
+            Owned<IDistributedSuperFile> superfile = dfd.createNewSuperFile(t);
+
+            superfile->validate();
```
I think there was an earlier comment , that validate would be better as a common method (virtual of IDistributedFile), for the super to do what you've implemented and for the base File version to do what it needs to do for file verification that you do below, i.e. if (file->existsPhysicalPartFiles(0)) ..

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+
+            try
+            {
+                superfile->attach(lfn, user, true);
```
attach is the right time to do something different re. subfiles, but it should have loaded them (and hence locked) them at constructor/init time as it is still doing.

See previous comment in last review:
>2) You do not want the init to loadSubFiles, because that will lookup and lock them and you are creating an unattached file (no Dali connection at this point). It should delay that action until attach().

createNewSuperFile() should avoid loading the supers.
attach() should load _and_ link them.

jakesmith replied:
This is still an issue afaics.
The original comment(s) still stand, a few reviews ago still stands:
```
2) You do not want the init to loadSubFiles, because that will lookup and lock them and you are creating an unattached file (no Dali connection at this point). It should delay that action until attach().
```

I see there are a 2 critical typo's in my last comment above:

>but it should have loaded them

Should have been:
>but it **shouldn't** have loaded them.

>createNewSuperFile() should avoid loading the supers.

Should have been:
>createNewSuperFile() should avoid loading the **subfiles.**


**You do not want an attached superfile holding locks..**

jakesmith replied:
Attaching a super (calling super->attach()) will always involve linking the subfiles, so having a different attach() prototype is not necessary afaics.
i.e. a single overloaded version of > void attach(const char *_logicalname,IUserDescriptor *user) 
is all that is required. It must load + link the subfiles (see previous comments above)

jakesmith replied:
My mistake, I was looking at the wrong constructor, I thought createNewSuperFile was calling the variety of CDistributedSuperFile that has the same 2 leading parameters, that calls loadSubFiles(), however, the one it calls is:
```
   CDistributedSuperFile(CDistributedFileDirectory *_parent, IPropertyTree *_root)
    {
        commonInit(_parent, _root);
    }
```

which does not.

jakesmith commented:
@AttilaVamos, getting there, some final comments/changes needed.

Please keep next commit as separate commit, thanks.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith updated, please re-review.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -4533,6 +4533,15 @@ protected: friend class CDistributedFilePart;
+        {
+            throw MakeStringException(-1, "Physical Parts doesn't exists for logical file !");
+        }
+
```
trivial: unnecessary extra newline here and a block with 1 line doesn't normally have open/close braces.
Also, exception text would be better as "Some physical parts do not exists, for logical file : %s", where logicalName was given if set and "[unattached]" was given if logicalName was not yet set.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5882,6 +5909,32 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+
+                if (!dfd.isSuperFile(subfilename.str(), NULL))
+                    if (!dfd.existsPhysical(subfilename.str(), NULL))
+                        throw MakeStringException(-1, "Physical Parts doesn't exists for logical file %s!", subfilename.str());
```
trivial: should read "Some physical parts do not exist for logical file %s!"

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5882,6 +5909,32 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+                subFileCount++;
+            }
+            if (numSubfiles != subFileCount)
+                throw MakeStringException(-1, "The value of @numsubfiles (%d) not equal with the number of SubFile items (%d)!",numSubfiles, subFileCount);
```
trivial: "The value of @numsubfiles (%d) is not equal to the number of SubFile items (%d)!"

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,56 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            catch (IException *e)
+            {
+                EXCLOG(e,"addFileXML(Super File");
+                e->Release();
```
I'm not convinced throwing the exception away is correct here..

AttilaVamos replied:
Re-throw?

jakesmith replied:
Yes, + check that it is correctly handled and reported by callers

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,56 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+
+            Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t, &queryNamedGroupStore(), 0);
+            Owned<IDistributedFile> file = dfd.createNew(fdesc, true);
+            if (file)
```
minor: pointless test, createNew cannot return NULL.

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,56 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                file->validate();
+                PROGLOG("Add logical file %s.", lfn);
+                file->attach(lfn,user);
+            }
```
With validate() and attach() being common virtuals (see previous comment re. single version of attach() prototype), and other comment re. exception handling, this code could be simplified to:

```
        Owned<IDistributedFile> file;
        if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_SuperFile)))
        {
            // It seems XML is a super file
            file.setown(dfd.createNewSuperFile(t));
        }
        else if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_File)))
        {
            // Logical file map
            Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t, &queryNamedGroupStore(), 0);
            file.setown(dfd.createNew(fdesc, true));
        }
        file->validate();
        PROGLOG("Adding %s file %s.", file->querySuperFile()?"super":"logical", lfn);
        file->attach(lfn,user);
```

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -432,11 +436,12 @@ interface IDistributedSuperFile: extends IDistributedFile
     virtual IDistributedFile *querySubPart(unsigned partidx,                // superfile part index
                                            unsigned &subfilepartidx)=0;     // part index in subfile
                                 // returns file for part (not linked) NULL if not found
+    virtual void validate() = 0;
```
Should be removed.

jakesmith commented on dali/base/dadfs.hpp:
```diff
@@ -409,6 +411,8 @@ interface IDistributedFile: extends IInterface
 interface IDistributedSuperFile: extends IDistributedFile
 {
+    virtual void attach(const char *logicalname,IUserDescriptor *user, bool link=false) = 0;     // attach to name in DFS
+
```
See previous comment re. overloaded std. attach() definition handling logic of loading and linking subfiles.
I don't think a special variety is needed.

jakesmith commented:
@AttilaVamos - some trivials comments, but 1 significant issue remains, regarding the loading/locking of subfiles before the super is attached.

AttilaVamos commented:
@jakesmith I updated accordingly your comments and our discuss. Please review.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,43 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+        const char * nodeName = t->queryName();
+        if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_SuperFile)))
+        {
+            // It seems XML is fine to create  super file
```
minor:

// It seems XML is a super file

[ from previous commit comment (https://github.com/hpcc-systems/HPCC-Platform/pull/7772#discussion_r47484327) ]

jakesmith commented on dali/dfu/dfuutil.cpp:
```diff
@@ -857,21 +857,43 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+            throw MakeStringException(-1, "Unrecognised file XML root tag detected: '%s'", nodeName);
+
+        file->validate();
+        PROGLOG("Add logical file %s.", lfn);
```
from previous commit comment also [https://github.com/hpcc-systems/HPCC-Platform/pull/7772#discussion_r47484327]

        PROGLOG("Adding %s file %s.", file->querySuperFile()?"super":"logical", lfn);

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -4533,6 +4533,15 @@ protected: friend class CDistributedFilePart;
+        if (!existsPhysicalPartFiles(0))
+        {
+            const char * logicalName = queryLogicalName();
+            throw MakeStringException(-1, "Some physical parts do not exists, for logical file : %s",(logicalName != "" ? logicalName : "[unattached]"));
```
> logicalName != ""
looks dodgy.
Should probably use : isEmptyString()

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5882,6 +5904,35 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+                    if (!dfd.existsPhysical(subfilename.str(), NULL))
+                    {
+                        const char * logicalName = queryLogicalName();
+                        throw MakeStringException(-1, "Some physical parts do not exists, for logical file : %s",(logicalName != "" ? logicalName : "[unattached]"));
```
Should use : isEmptyString()

jakesmith commented:
@AttilaVamos - back to you.

AttilaVamos commented:
@jakesmith updated. Please check it again.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -5882,6 +5904,34 @@ class CDistributedSuperFile: public CDistributedFileBase<IDistributedSuperFile>
+                if (!parent->exists(subfilename.str(), NULL))
+                    throw MakeStringException(-1, "Logical subfile '%s' doesn't exists!", subfilename.str());
+
+                if (!parent->isSuperFile(subfilename.str(), NULL, NULL, 0))
```
trivial: you could argue that adding the default params in the implementation of isSuperFile() to mirror the interface method definition would be better than adding extra ,NULL, 0 here.

jakesmith commented:
Apart from trivial comment looks good to merge to me.
@AttilaVamos - can you squash changes into 1 commit now ?

AttilaVamos committed:
HPCC-14111 To dfuplus, add capability to export and restore super files.

Improve savexml action to generate proper logical and super file xml blob.

Improve add action to load simple logical and superfile metainformation
from an XML file. If the restored metainformation belong to a superfile
then update all sub-files SuperOwner tag.

Fix missing '\n' at the end of some DFUPlus return messages.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith updated accordingly your last comment.

AttilaVamos commented:
@richardkchapman or @ghalliday please merge

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
